### DISCUSSION TITLE
Please vote in the poll: Liberate all proprietary software on which these modules depend

### DISCUSSION THREAD

**Discussion by throwaway1037** [NVIDIA/open-gpu-kernel-modules#292] [Polls]
As discussed in: #291 these modules depend on proprietary software.
These are:
- `gsp.bin`
- Those in the [Linux-firmware repo](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/nvidia).
See also: #277.
- I recently discovered there are proprietary binary blobs included in source files: #286.
- The userspace binaries from the proprietary driver.

The driver requires all of these components to work.

Without the firmware from Linux-firmware, core features like reclocking will not function, which hindered Nouveau for years.
It's also (virtually) impossible to replace them by reverse-engineering because they are signed and the GPU will refuse to load firmware which is not signed by nVidia.

OS distributions such as [Parabola GNU+Linux-libre](https://parabola.nu/) use [Linux-libre](https://linux-libre.fsfla.org), a version of Linux without blobs, and does not distribute blobs.
In essence, without all the software on which these modules depend also being free software, this awesome free software driver release is crippled in the free world, as it relies on proprietary binaries.

I started this poll at the recommendation of @PAR2020 in: #291.

If you want to send the message to NVIDIA that we want our freedom, please vote "Yes" in the poll.
Thank you!

**Comment by YusufKhan-gamedev** [comment]
I voted yes but thats NOT THE POINT!

If you wish you can disassemble the blobs(that you can! Some blobs are illegal to disassemble) but AMD and intel and Qualcomm etc ALL HAVE PROPRIETARY FIRMWARE your cpu has proprietary firmware, run `sudo rm -rf /lib/firmware` if you wish to remove them! Be warned your computer might brick!

**Reply by throwaway1037** [reply]
Use [Linux-libre-firmware](https://jxself.org/firmware) if you want all your firmware to be libre.

It is also worth noting that the chances of your computer becoming bricked from removing the proprietary firmware is extremely small; the vast majority of computers will boot without blobs.

**Reply by YusufKhan-gamedev** [reply]
The reason for the computer booting is that firmware still exists in your cpu, if you remove it(somehow) your computer will brick.

**Reply by throwaway1037** [reply]
Firmware physically encoded inside the CPU is hardware, not software.

**Reply by YusufKhan-gamedev** [reply]
In that case my Nintendo 3ds is a free software Richard Stallman Approved(tm) device!

**Reply by throwaway1037** [reply]
No, because it contains proprietary software.
You're conflating hardware with software.

**Reply by YusufKhan-gamedev** [reply]
Hahahah "Firmware physically encoded inside the {device} is hardware, not software."  isnt it? My firmware is physically encoded onto my 3ds and gets updates just like the microcode and spyware in our cpus! No difference noted.

**Reply by throwaway1037** [reply]
As I said, 3DSs contain proprietary software.
You're conflating hardware (immutable circuits which physically encode immutable data) with software (mutable data).

**Reply by YusufKhan-gamedev** [reply]
CPU microcode is mutable you can update it....

**Reply by throwaway1037** [reply]
No, the microcode is burnt into ROM.
All the microcode update bundles that exist at the moment are proprietary software.

**Reply by DemiMarie** [reply]
@throwaway1037 you are missing the point: on x86 the only options are either more bugs (built-in microcode) or fewer bugs (updated microcode).  If you want a fully libre system your only option is POWER9 but that is very expensive.

**Reply by aviallon** [reply]
In my country it is never illegal to disassemble firmware blobs, so I do not care.

**Reply by arisu3** [reply]
Microcode updates are more like complex, encrypted configuration files. They aren't directly executable and they can't make arbitrary changes to the CPU's behavior. Microcode is only software in the loosest sense.

**Comment by Evernow** [comment]
There is no benefit to the consumer for this. There hasn't been any hardware that has libre firmware since 4th gen Intel.

**Reply by throwaway1037** [reply]
There is benefit to everybody, as the firmware should respect their freedom.
As I stated in the original post, people who want to keep their freedom by using a 100% libre OS, cannot use NVIDIA cards, at least not fully without the firmware being libre.
The fact that there is a lack of libre firmware is a statement of the problem, not a justification of restricting people and not respecting their freedom.

**Reply by YusufKhan-gamedev** [reply]
> cannot use NVIDIA cards

Ehh the firmware for nouveau is free*...you can use it...it supports reclocking*....

**Reply by throwaway1037** [reply]
No, the firmware is proprietary.
That's the problem.

To quote your own [Github profile](https://github.com/YusufKhan-gamedev)
> I use my github accounts mainly for shitposting and other worthless work.

I am treating your previous conflations of software and hardware in another comment on this discussion and the reply you wrote on this comment as such.

**Reply by YusufKhan-gamedev** [reply]
https://github.com/teslamotors/linux/tree/tegra-4.4/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/fuc

This is re-clocking firmware, free, its not hard to find it...Im more inclined that your account is more of a troll than mine since all you talk about is getting software under a free licence...on github for that matter....

**Reply by throwaway1037** [reply]
1. [This file](https://github.com/teslamotors/linux/blob/tegra-4.4/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/fuc/gf119.fuc4.h) for instance is just binary blobs disguised as source code in the form of long hex arrays. If they are built from the source code files in the repo, then they should be done so at build-time instead of being hardcoded, otherwise it's difficult to verify they are fully libre, correspond to the source files, and have no proprietary software added in.
2. Nouveau relies on proprietary firmware in the Linux-firmware repo.
3. That firmware is for Tegra, which is irrelevant to these kernel modules.

**Reply by YusufKhan-gamedev** [reply]
1. They are telling you what the blobs are doing, they are blobs because there is no reason to create a special source form for it if they arent going to be actually using the source form(see the comments in the "blob")
2. Yes, but there are some chipsets that dont work with that firmware, those are the ones with support.
3. They include both tegra and desktop GPU chipsets, example https://www.techpowerup.com/gpu-specs/nvidia-gk208.g572 the repository I shows was for teslas linux kernel for tegra cars since that was open at the time of me writing this

**Reply by princesstwi** [reply]
Still different. Kepler is outdated and has nothing to do with these modules. The reclocking firmware for Ampere and Turing is still closed source. I and many others don't even own a Kepler  (gt/gtx 6xx to 7xx). The 9xx seris (Maxwell)  and 10xx seris Pascal are the oldest still usable desktop cards. Turing is 16xx to 20xx. Ampere is 30xx. (The ones that use gsp)

**Reply by throwaway1037** [reply]
> 1. They are telling you what the blobs are doing, they are blobs because there is no reason to create a special source form for it if they arent going to be actually using the source form(see the comments in the "blob")

Irrelevant.
If there is no source, they are proprietary.

> 2. Yes, but there are some chipsets that dont work with that firmware, those are the ones with support.

Regardless, Nouveau depends on proprietary firmware to support most cards.

**Reply by YusufKhan-gamedev** [reply]
1. If you consider that proprietary then you should consider gcc as proprietary because it includes opcodes without the source code to the cpu...linux-libre contains these "blobs" because they are the preferred way that redhat modified the firmware, if you want you can fork linux-libre with a version that doesnt contain those blobs.
2.  Nouveau depends on non-free blobs for most cards to get better 3D performance, its not needed.

@princesstwi This thread is about freedom, all of that firmware are for the cards that support re-clocking since creating new firmware would not matter as they already have firmware even though its not good firmware nothing would have been improved(in redhats eyes) if they created their own firmware, also 4xx-7xx is still more than usable for gaming...ive only heard 9xx+ whine about performance.

**Reply by throwaway1037** [reply]
> 1. If you consider that proprietary then you should consider gcc as proprietary because it includes opcodes without the source code to the cpu...

What do you mean?
What is an "opcode without the source to the CPU"?
That's grammatically incorrect.

> 2. Nouveau depends on non-free blobs for most cards to get better 3D performance, its not needed.

Some cards require NVIDIA blobs for NVENC, etc.
Regardless, even if they were optional, the problem is that the blobs are proprietary.

> @princesstwi This thread is about freedom, all of that firmware are for the cards that support re-clocking since creating new firmware would not matter as they already have firmware even though its not good firmware nothing would have been improved(in redhats eyes) if they created their own firmware, also 4xx-7xx is still more than usable for gaming...ive only heard 9xx+ whine about performance.

This is also incoherent and grammatically incorrect.
Those firmwares you linked to have nothing to do with these modules.

Also I am now sure you're a troll.
It can't be possible for someone to know so much about tech but get so much wrong.

**Reply by YusufKhan-gamedev** [reply]
1. My point was that there is no source code to what the opcodes are in a CPU just like there is no source code for the "blobs" in the free nouveau firmware.

> Those firmwares you linked to have nothing to do with these modules.

You complained there wasnt a way to use NVIDIA with free firmware, I showed you it, you said it wasnt free even though linux-libre includes them.

> It can't be possible for someone to know so much about tech but get so much wrong.

Same to you Mr. throwaway, im not spending much time on this thread because your a troll(hence my grammer).

**Reply by throwaway1037** [reply]
> 1. My point was that there is no source code to what the opcodes are in a CPU just like there is no source code for the "blobs" in the free nouveau firmware.

You're once again conflating hardware and software.

> > Those firmwares you linked to have nothing to do with these modules.
> 
> You complained there wasnt a way to use NVIDIA with free firmware, I showed you it, you said it wasnt free even though linux-libre includes them.

Once again you're missing the point.
It doesn't matter if there is a minimal amount of supposedly free firmware for Tegra, the issue is that proprietary firmware is required to use the cards we are discussing.
If Linux-libre includes those Tegra blobs, then either they are actually libre or you should report this as a bug.
I never claimed they were proprietary, I said they are proprietary if there is no source and that the best way to have the repo set up is for them to be built at build-time instead of hardcoded as hex arrays.

**Reply by YusufKhan-gamedev** [reply]
1. No, your conflating hardware and software, the "blobs" are as much peices of hardware as opcodes are, in fact they are *kinda* opcodes.
2. Just because it works on tegra doesnt mean it only works on tegra, it works on 4xx-7xx GPUs, and you did claim they were proprietary:

> [This file](https://github.com/teslamotors/linux/blob/tegra-4.4/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/fuc/gf119.fuc4.h) for instance is just binary blobs disguised as source code in the form of long hex arrays

> No, the firmware is proprietary.

**Reply by throwaway1037** [reply]
> 1. No, your conflating hardware and software, the "blobs" are as much peices of hardware as opcodes are, in fact they are _kinda_ opcodes.

Software blobs are not hardware.

> 2. Just because it works on tegra doesnt mean it only works on tegra, it works on 4xx-7xx GPUs,

Which are irrelevant to this discussion, which concerns cards that use these modules.

> and you did claim they were proprietary:
> 
> 
> > [This file](https://github.com/teslamotors/linux/blob/tegra-4.4/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/fuc/gf119.fuc4.h) for instance is just binary blobs disguised as source code in the form of long hex arrays

They may have been generated from libre source code and as such are libre.

>
> > No, the firmware is proprietary.

That was referring to the Linux-firmware repo blobs.

I'm tired of correcting your misinformation and trolling, so I'm not going to reply to any more of your nonsense.
Go away.

**Reply by YusufKhan-gamedev** [reply]
The libre firmware found in linux-libre and linux(the source code, not the firmware repo) is 100% free, the hex arrays are not proprietary as they are hardcoded onto the GPU and do not have a source form so they are hardware(in your definition). You can use nvidia gpus with nouveau without firmware anyway, you just wont get the absolute best performance. If you wish, remove firmware loading from this repo and fork it, otherwise just use nouveau with linux-libre(If you have a nouveau card).

**Comment by princesstwi** [comment]
I voted yes even though i don't think this poll will be convencing enough for nvidia. Like the other comments say, not even intel or amd open source their firmware.
What i want to see most of all is open source userspace so the mesa team can develop a more powetful nouveau

**Reply by throwaway1037** [reply]
That is in the fourth bullet point.
See also [why "open source" misses the point](https://gnu.org/philosophy/open-source-misses-the-point.html).

**Comment by princesstwi** [comment]
Just pointing out tegra isn't for desktop. It is for arm devices like switch and tablets. If anyone else reads this please disreguard anyone who says tegra is the same.
"Cough, tesla, cough."

**Reply by YusufKhan-gamedev** [reply]
The firmware I showed was not just for tegra, it was for all nvidia devices 4xx-7xx and its mainline

**Comment by ghost** [comment]


**Reply by YusufKhan-gamedev** [reply]
Or at least provide a stable firmware ABI and document it

**Comment by 64kramsystem** [comment]
Vote with your wallet (I did), and buy AMD if you don't like Nvidia's model.

**Comment by arisu3** [comment]
I understand the purpose of keeping the GSP closed. Even according to the strictest definitions of free software, binary blobs are not necessarily problematic as they are not running on your system.

But the closed userspace components should absolutely be released. I also think the GSP firmware should be released, but I think there is a 0% chance of that happening. I think there is a non-zero chance of userspace components being opened one day, which is why I voted yes.
